server/docs/design/block-persistence.md

```diff
@@ -3,64 +3,213 @@
+  <dt>BlockWriter</dt>
+  <dd>An interface defining methods used to write `Blocks` to storage. It
+  represents a lower-level component whose implementation is directly
+  responsible for writing a `Block` to storage.</dd>
 
-**BlockReader** - An interface defining methods used to read a Block from storage. It represents a lower-level
-component whose implementation is directly responsible for reading a Block from storage.
+  <dt>BlockRemover</dt>
+  <dd>An interface defining the methods used to remove a `Block` from storage.
+  It represents a lower-level component whose implementation is directly
+  responsible for removing a `Block` from storage.</dd>
 
-**BlockWriter** - An interface defining methods used to write BlockItems to storage. It represents a lower-level
-component whose implementation is directly responsible for writing a BlockItem to storage as a Block.
+  <dt>BlockPathResolver</dt>
+  <dd>An interface defining methods used to resolve the path to a `Block` in
+  storage. It represents a lower-level component whose implementation is
+  directly responsible for resolving the path to a `Block` in storage, based on
+  `StorageType`.</dd>
+</dl>
+
+## Goals
 
-**BlockRemover** - An interface defining the methods used to remove a Block from storage. It represents a lower-level
-component whose implementation is directly responsible for removing a Block from storage.
+1. `BlockItems` streamed from a producer (e.g. Consensus Node) must be collated
+   and persisted as a `Block`. Per the specification, a `Block` is an ordered
+   list of `BlockItems`. How the `Block` is persisted is an implementation 
+   detail.
+1. A `Block` must be efficiently retrieved by block number.
+1. Certain aspects of the `Block` persistence implementation must be
+   configurable.
+
+## Overview
+
+The design for `Block` persistence is fairly straightforward. `Block` server 
+objects should use the persistence abstractions to read, write and remove 
+`Block`s from storage, as well as resolve the paths to `Block`s. `BlockItem`s
+streamed from a producer  are read off the wire one by one and passed to an
+implementation of `BlockWriter`. The `BlockWriter` is responsible for collecting
+related `BlockItem`s into a `Block` and persisting the `Block` to storage in a
+way that  is efficient for both long-term storage and rapid retrieval at a later
+time. The `BlockWriter` is also  responsible for removing a partially written
+`Block` if an exception occurs  while writing it. For example, if half the
+`BlockItem`s of a `Block` are written  when an `IOException` occurs, the
+`BlockWriter`should remove all the `BlockItem`s of the partially written `Block`
+and pass the exception up to the  caller. Services requiring one or more `Block`
+s should leverage a `BlockReader` implementation. The `BlockReader` should be
+able to efficiently retrieve a `Block` by block number.  The `BlockReader`
+should pass unrecoverable exceptions  when reading a `Block` up to the caller.
```

Comment:
Adding the `s` for `Block`s at the start of the next line adds an unintended space.
Sometimes the automated formatters need a little help.
Also, "leverage" is heavily overused; so it can be helpful to replace with "use"
here.
```suggestion
and pass the exception up to the  caller. Services requiring one or more
`Block`s should use a `BlockReader` implementation. The `BlockReader` will be
able to efficiently retrieve a `Block` by block number.  The `BlockReader`
will pass unrecoverable exceptions when reading a `Block` up to the caller.
```

Reply:
✅ Done.